### ISSUE TITLE
All traffic not using proxy when enabled (2.0.19)

### ISSUE THREAD

**Issue by desultory-zz** [DNSCrypt/dnscrypt-proxy#805]
I enabled the socks proxy support using a tor proxy and dnscrypt-proxy is still sending traffic outside of it.

I am using the config lines:

```
dnscrypt_servers = true
doh_servers = false
force_tcp = true
proxy = "socks5://127.0.0.1:9050"
```

as recommended in the configuration file, but I ran a packet capture on my network and there is still traffic being sent directly to dns servers which is plaintext as well.  I was able to resolve this by adding the following rules to iptables:

```
iptables -A OUTPUT -s 127.0.0.0/9 -j ACCEPT
iptables -A OUTPUT --proto UDP --dport 53 -j DROP
```
and dnscrypt still works but all unencrypted outbound traffic is dropped.

There is no indication that when using a socks proxy, all traffic is not sent through it, and I even disabled the doh_servers directive because the configuration example stated that a http proxy is required for that.

I am also running a bind9 server on this machine that is the primary dns server, but its only forward lookup is 127.0.0.2:53 where I have dnscrypt-proxy bound to. When testing this and performing the packet capture, I ran:
`
dnscrypt-proxy -config /etc/dnscrypt-proxy/dnscrypt-proxy.toml -resolve {domain}
`
to ensure all traffic was going through dnscrypt-proxy and not somehow going to an incorrect dns server through bind.

**Comment by jedisct1** [comment]
I guess what you are seeing is the normal query to retrieve certificates. There is nothing confidential here, it's all public information. And that data is signed, so it cannot be modified without this being detected.

**Comment by desultory-zz** [comment]
I was not seeing those queries.  I had my fallback DNS server to use 1.1.1.1 and the queries I logged with wireshark were intermittent and matched the test URL I used.  I tried several domains I knew I had not resolved recently to ensure I got fresh results in addition to disabling caching.  In the wireshark log, I could see mostly TOR traffic coming from that IP but there were also plaintext DNS requests which contained the requests for the domains I was testing.

**Comment by desultory-zz** [comment]
![Untitled](https://user-images.githubusercontent.com/11171762/56944890-0eeac000-6b1d-11e9-8259-c34fe815ee38.png)
I could be wrong, but I'd expect the proxy option to push all traffic through the proxy, as you can see here, a substantial amount is leaking

In this case, the TOR node I was connected to was "23.81.66.90", I'd expect nearly all traffic originating from that server to go through that, given that this is a gentoo based server and the packet capture was done on the WAN interface so all internal traffic was ignored.

**Comment by jedisct1** [comment]
These queries are not sent by dnscrypt-proxy.

**Comment by desultory-zz** [comment]
Those queries were the result of me running "dnscrypt-proxy -config /etc/dnscrypt-proxy/dnscrypt-proxy.toml -resolve {domain}" and then checking for those domains in the packet capture.  I'm not sure what else could have been making those requests.  I tried this with a handful of domains, and if I ran the command several times in a row, eventually one would appear like that. 

My server configuration is as follows: 
Server is set to use 127.0.01 as the primary DNS resolver
Bind9 is running on 127.0.0.1:53 and has only one resolver as the forward lookup resolver which is 127.0.0.2
dnscrypt-proxy is running on 127.0.0.2:53 and is set to push all traffic through my TOR socks5 proxy on 127.0.0.1:9050
With this configuration, there should be no way any traffic leaves through 10.200.0.53 to an public DNS server unless a service is deliberately doing that which I don't believe is the case given that I built this server from scratch using gentoo.

**Comment by jedisct1** [comment]
dnscrypt-proxy doesn't have code to send unencrypted queries, except in the forwarding module.

>  Server is set to use 127.0.01 as the primary DNS resolver

Is it the ONLY configured resolver?

**Comment by desultory-zz** [comment]
This is the resolv.conf on the server: 
```
domain holmium.mylocal
nameserver 127.0.0.1
```
This is the Bind9 Configuration:
```
acl "xfer" {
        none;
};

acl "trusted" {
        127.0.0.0/8;
        10.200.0.0/24;
        10.8.0.0/24;
};

options {
        directory "/var/bind";
        pid-file "/run/named/named.pid";

        response-policy {
                zone "holmium.mylocal";
                zone "rpz";
                zone "blacklist";
        };

        listen-on {
                127.0.0.1;
                any;
        };

        allow-query {
                trusted;
        };

        allow-query-cache {
                trusted;
        };

        allow-recursion {
                trusted;
        };

        allow-transfer {
                none;
        };

        allow-update {
                none;
        };

        forward first;
        forwarders { 127.0.0.2; };

        dnssec-enable yes;
        dnssec-validation auto;
        dnssec-must-be-secure holmium.mylocal no;
        dnssec-must-be-secure rpz no;
        dnssec-must-be-secure blacklist no;
        //query-source address * port 53;
};

```

And this is the dnscrypt-proxy configuration:
```
## Online documentation is available here: https://dnscrypt.info/doc
#server_names = ['cloudflare']
listen_addresses = ['127.0.0.2:53']
max_clients = 500
# user_name = 'nobody'
ipv4_servers = true
ipv6_servers = false
dnscrypt_servers = true
doh_servers = false
require_dnssec = true
require_nolog = true
require_nofilter = true
force_tcp = true
proxy = "socks5://127.0.0.1:9050"
timeout = 5000
keepalive = 30
#refused_code_in_responses = false
cert_refresh_delay = 240
dnscrypt_ephemeral_keys = true
tls_disable_session_tickets = true
tls_cipher_suite = [52393, 52393]

fallback_resolver = '1.1.1.1:53'
ignore_system_dns = false
netprobe_timeout = 60
# offline_mode = false
log_files_max_size = 1
log_files_max_age = 1
log_files_max_backups = 1

block_ipv6 = true
cache = false
#cache_size = 512
#cache_min_ttl = 600
#cache_max_ttl = 86400
#cache_neg_min_ttl = 60
#cache_neg_max_ttl = 600

[sources]
  [sources.'public-resolvers']
  urls = ['https://raw.githubusercontent.com/DNSCrypt/dnscrypt-resolvers/master/v2/public-resolvers.md', 'https://download.dnscrypt.info/resolvers-list/v2/public-resolvers.md']
  cache_file = '/var/cache/dnscrypt-proxy/public-resolvers.md'
  minisign_key = 'RWQf6LRCGA9i53mlYecO4IzT51TGPpvWucNSCh1CBM0QTaLn73Y7GFO3'
  refresh_delay = 72
  prefix = ''

```


I have ipv6 disabled in the kernel so I know nothing is "sneaking by" using ipv6 or anything but as you can see, the system is told to use 127.0.0.1, bind9 is using 127.0.0.1:53, bind9 is told to ONLY use its internal zones and cache and then to try to use 127.0.0.2(dnscrypt-proxy) if it needs to look something up, and then in dnscrypt-proxy I have caching disabled(mostly for testing but I also think 2 caches is redundant), and I even have dnscrypr set to not use DOH in case that didn't pass through the socks proxy.
What makes me almost certain that the requests must be passing through dnscrpyt proxy is the fact that with this configuration, it should be using the fastest server from the resolver lists and I have never once configured my server to use and resolver but 1,1.1.1 manually.  I would expect any unencrypted request to go to 1.1.1.1 because that is what my dhcp server has set as the fallback one and I also set the resolv.conf to be immutable in this case so it shouldn't even be able to set that, even if it did get a DHCP update.
The unencrypted requests that it did make went through some random resolver that I didn't recognize so I am nearly certain that it went through dnscrypt proxy and not something else.  I did a test with the domain "brazzers.com" just to be certain that it was dnscrypt doing the resolving and got this:
![Untitled2](https://user-images.githubusercontent.com/11171762/56969962-15e5f280-6b5e-11e9-967e-d5b31e640f5d.png)
I am 100% sure that request was made using: 
`dnscrypt-proxy -config /etc/dnscrypt-proxy/dnscrypt-proxy.toml -resolve brazzers.com`
because I never visit that site so this would be the only reason that was requested.  While testing this, I was testing with a friend who technically owns the network so we were just trying to confirm that this was happening (because it should not have been happening) and started running tests using domains that we knew had to have been through dnscrypt and not some other service.
I also tested this using dig @127.0.0.2 {domain} but I took those tests with a grain of salt because I believe there is a remote possibility that using that command could result in a leak, but using the dnscrypt command should not be able to use any resolver but dnscrypt theoretically.

**Comment by jedisct1** [comment]
Try putting Bind out of the equation. Stop it and have dnscrypt-proxy listen to `127.0.0.1`.

**Comment by desultory-zz** [comment]
Just attempted it again, disabled selinux and iptables, and for some reason I'm getting: 
```
dnscrypt-proxy -config dnscrypt-proxy.toml -resolve google.com
Resolving [google.com]

Domain exists:  probably not, or blocked by the proxy
Canonical name: -
IP addresses:   -
TXT records:    -
```
I just woke up, and my configuration is entirely unchanged.  iptables is not blocking any requests, the firewall is not blocking any requests, selinux is not doing anything but auditing and for some reason that command is no longer working. dig @127.0.0.2 google.com does work for some reason though but this entire situation is becoming more confusing.  I still ran the packet capture and there is a lot of plain dns traffic leaving the system, presumably through dnscrypt because at the time, bind9 was not running.  I'm going to have to do a lot more debugging since this new issue has seemingly come out of nowhere.
